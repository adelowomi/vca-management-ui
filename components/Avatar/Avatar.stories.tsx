import { number, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';

import Avatar from './Avatar';

storiesOf('Avatar', module)
  .addDecorator(withKnobs)
  .add('Avatar Demo', () => (
    <Avatar
      firstName={text('First name', 'John')}
      lastName={text('Last name', 'Doe')}
      size={number('Progress', 18, { range: true, min: 0, max: 100 })}
    />
  ))
  .add('Avatar without initials', () => (
    <Avatar
      firstName={text('First name', 'John')}
      lastName={text('Last name', 'Doe')}
      hasInitials={false}
      size={number('Progress', 30, { range: true, min: 0, max: 100 })}
    />
  ));
