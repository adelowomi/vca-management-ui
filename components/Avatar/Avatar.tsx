import React, { useMemo } from 'react';

import mergeClassNames from '../../helpers/mergeClassNames';
import { User } from '../AssetsSVG';

const FONT_SIZE_LIMIT = {
  lower: 18,
  upper: 24,
};

interface AvatarProp {
  /**
   * To provide custom styles.
   */
  className?: string;
  /**
   * Adds first name which would be used to generate initials.
   */
  firstName?: string;
  /**
   * Would render initials if set to true. Default is true.
   */
  hasInitials?: boolean;
  /**
   * Adds last name which would be used to generate initials.
   */
  lastName?: string;
  /**
   * Set the size of the icon. Default is 18.
   */
  size?: number;
}

const Avatar: React.FC<AvatarProp> = ({
  className,
  firstName,
  hasInitials,
  lastName,
  size,
}) => {
  /** Generates initials based on the first letters of the name and surname
   */
  const initials = useMemo(() => {
    if (!firstName || !lastName) {
      return null;
    }

    const firstNameInitial = firstName.charAt(0);
    const lastNameInitial = lastName.charAt(0);

    return size < FONT_SIZE_LIMIT.upper
      ? firstNameInitial
      : `${firstNameInitial}${lastNameInitial}`;
  }, [firstName, lastName, size]);

  /** Returns width and height of the avatar's container */
  const containerStyle = useMemo(
    () => ({
      height: size,
      width: size,
    }),
    [size]
  );

  /** Returns styles for initials */
  const initialsStyles = useMemo(
    () => ({
      lineHeight: `${size}px`,
      fontSize: `${size >= FONT_SIZE_LIMIT.upper ? size / 3 : size / 2}px`,
    }),
    [size]
  );

  /** Returns boolean value determining if initials should be displayed
   * @returns {boolean}
   */
  const shouldDisplayInitials = useMemo(
    () => hasInitials && size > FONT_SIZE_LIMIT.lower,
    [hasInitials, size]
  );

  const renderedInitials = useMemo(
    () =>
      shouldDisplayInitials && (
        <span
          className={`absolute top-0 left-0 w-full h-full text-center app-white ${initialsStyles}`}
          data-testid="initials-test"
          style={initialsStyles}
        >
          {initials}
        </span>
      ),
    [initials, initialsStyles, shouldDisplayInitials]
  );

  return (
    <div
      className={mergeClassNames('relative', className)}
      style={containerStyle}
      data-testid="avatar-test"
    >
      <User size={size} className="app-avatar-icon" />
      {renderedInitials}
    </div>
  );
};

Avatar.defaultProps = {
  className: '',
  firstName: '',
  hasInitials: true,
  lastName: '',
  size: FONT_SIZE_LIMIT.lower,
};

export default Avatar;
