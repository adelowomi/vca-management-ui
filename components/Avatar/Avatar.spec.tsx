import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import React from 'react';

// Components
import Avatar from './Avatar';

describe('Avatar', () => {
  describe('render', () => {
    describe.each([
      ['Femi', 'Adeojo', 32, true],
      ['James', 'Bond', 32, false],
    ])(
      'with firstname %s lastname %s size %s and hasInitials %s set to',
      (firstName, lastName, size, hasInitials) => {
        it('render status of shipment', () => {
          const { container } = render(
            <Avatar
              firstName={firstName}
              lastName={lastName}
              size={size}
              hasInitials={hasInitials}
            />
          );
          expect(container.firstChild).toMatchSnapshot();
        });
      }
    );
  });

  describe('with props provided', () => {
    it('should set size based on size prop', () => {
      const { getByTestId } = render(
        <Avatar firstName="John" lastName="Doe" size={64} />
      );
      expect(getByTestId('avatar-test')).toHaveStyle({ width: '64px' });
    });

    it('should override class', () => {
      const { getByTestId } = render(
        <Avatar className="test" firstName="John" lastName="Doe" size={64} />
      );
      expect(getByTestId('avatar-test')).toHaveAttribute(
        'class',
        'relative test'
      );
    });

    it('should render initials based on firstName and lastName props', () => {
      const { getByText } = render(
        <Avatar className="test" firstName="John" lastName="Doe" size={64} />
      );
      expect(getByText('JD')).toBeInTheDocument();
    });
  });

  describe('when size < 24px', () => {
    it('should show single initial', () => {
      const { queryByText, queryByTestId } = render(
        <Avatar className="test" firstName="John" lastName="Doe" size={20} />
      );
      expect(queryByText('J')).toBeInTheDocument();
      expect(queryByTestId('initials-test').textContent.trim().length).toBe(1);
    });
  });

  describe('when size < 18px', () => {
    it('should not show initial', () => {
      const { queryByTestId } = render(
        <Avatar className="test" firstName="John" lastName="Doe" size={12} />
      );
      expect(queryByTestId('initials-test')).toBeNull();
    });
  });

  describe('when firstName prop is not passed', () => {
    it('should not render Avatar with initials', () => {
      const { queryByTestId } = render(<Avatar size={30} />);
      expect(queryByTestId('initials-test')).toHaveTextContent('');
    });
  });
});
