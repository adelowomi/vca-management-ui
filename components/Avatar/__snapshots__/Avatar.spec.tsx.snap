// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar render with firstname Femi lastname Adeojo size 32 and hasInitials true set to render status of shipment 1`] = `
<div
  class="relative"
  data-testid="avatar-test"
  style="height: 32px; width: 32px;"
>
  <div
    class="app-icon app-avatar-icon"
    style="width: 32px; height: 32px;"
  >
    <i>
      <svg
        height="1000px"
        viewBox="0 0 96 96"
        width="1000px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="96"
          width="96"
        />
        <g
          fill="#fff"
          style="opacity: 0.15;"
        >
          <path
            d="M68.25,41.25c-.61,11.6-8.47,21-20.25,21s-19.64-9.46-20.25-21L27,27.75c0-10.31,7.5-21,21-21,5,0,13.5,1.59,14.25,8.25,3.66.09,4.3,2.24,5.57,4.82S69,25,69,27.75Z"
          />
          <path
            d="M55.5,67.5l-7.5,9-7.5-9C24.66,69.37,8.22,76.63,3.56,83.73A7.83,7.83,0,0,0,2.25,87c0,.48-.44,4.18-.75,9h93c-.31-4.77-.75-8.62-.75-9a5.54,5.54,0,0,0-1.31-3.25C87.79,76.63,71.34,69.37,55.5,67.5Z"
          />
        </g>
      </svg>
    </i>
  </div>
  <span
    class="absolute top-0 left-0 w-full h-full text-center app-white [object Object]"
    data-testid="initials-test"
    style="line-height: 32px; font-size: 10.666666666666666px;"
  >
    FA
  </span>
</div>
`;

exports[`Avatar render with firstname James lastname Bond size 32 and hasInitials false set to render status of shipment 1`] = `
<div
  class="relative"
  data-testid="avatar-test"
  style="height: 32px; width: 32px;"
>
  <div
    class="app-icon app-avatar-icon"
    style="width: 32px; height: 32px;"
  >
    <i>
      <svg
        height="1000px"
        viewBox="0 0 96 96"
        width="1000px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="96"
          width="96"
        />
        <g
          fill="#fff"
          style="opacity: 0.15;"
        >
          <path
            d="M68.25,41.25c-.61,11.6-8.47,21-20.25,21s-19.64-9.46-20.25-21L27,27.75c0-10.31,7.5-21,21-21,5,0,13.5,1.59,14.25,8.25,3.66.09,4.3,2.24,5.57,4.82S69,25,69,27.75Z"
          />
          <path
            d="M55.5,67.5l-7.5,9-7.5-9C24.66,69.37,8.22,76.63,3.56,83.73A7.83,7.83,0,0,0,2.25,87c0,.48-.44,4.18-.75,9h93c-.31-4.77-.75-8.62-.75-9a5.54,5.54,0,0,0-1.31-3.25C87.79,76.63,71.34,69.37,55.5,67.5Z"
          />
        </g>
      </svg>
    </i>
  </div>
</div>
`;
